### DISCUSSION TITLE
Is Patroni removing the leader lock too early on failover request?

### DISCUSSION THREAD

**Discussion by klaus-nicat** [patroni/patroni#3235] [General]
Current Leader db2: manual migration to db1:

```
Current Leader db2: manual migration to db1:
12:14:29 db2 patroni[1845]: 2024-12-11 12:14:29,036 INFO: received failover request with leader=None candidate=db1.rcode0.net scheduled_at=None
12:14:29 db2 patroni[1845]: 2024-12-11 12:14:29,367 INFO: Demoting self (graceful)
12:14:31 db2 patroni[3371]: patroni_postgres: 2024-12-11 12:14:31.756 UTC 3371 LOG:  received fast shutdown request
12:14:31 db2 patroni[3371]: patroni_postgres: 2024-12-11 12:14:31.758 UTC 3371 LOG:  aborting any active transactions
12:14:32 db2 patroni[3772]: patroni_postgres: 2024-12-11 12:14:32.080 UTC 3772 LOG:  shutting down
12:14:38 db2 patroni[1845]: 2024-12-11 12:14:38,203 INFO: updated leader lock during manual failover: demote
12:14:52 db2 vip-manager[1854]: 2024/12/11 12:14:52 current leader from DCS:
12:14:52 db2 vip-manager[1854]: 2024/12/11 12:14:52 IP address xx.xx.106.10/24 state is true, desired false
12:14:52 db2 vip-manager[1854]: 2024/12/11 12:14:52 Removing address xx.xx.106.10/24 on enp1s0
12:14:57 db2 patroni[58038]: patroni_postgres: 2024-12-11 12:14:57.380 UTC 58038 LOG:  database system was shut down at 2024-12-11 12:14:51 UTC
12:14:57 db2 patroni[58038]: patroni_postgres: 2024-12-11 12:14:57.380 UTC 58038 LOG:  entering standby mode
```

As you see, Patroni initiates the Postgresql shutdown, it demotes the leader lock, and some time later the Postgresql shutdown finishes. We are using VIP-manager to add/delete the VIP from the servers. In the above case, vip-manager detects that the leader lock was removed and removes the VIP from the server.

This caused problems for Postgresql, as some of the replication clients still had open TCP connections. Hence, the TCP connections could not be terminated properly and several of our Postgresql clients had broken TCP connections and were in dangling state. We had to restart the clients to recover (we waited > 15 minutes for self-healing, but then we restarted. Mabye the would have recovered after the default Linux TCP keepalive teimout of 2h).

So, wouldn't it be better if Patroni only removes the leader lock when Postgresql finished the shutdown?

**Comment by CyberDem0n** [comment]
Patroni indeed removes the leader lock while postgres is still running, but there are two preconditions for it:
1. pg_control file already reports `Database cluster state` as `shut down`.
2. It is known that at least one of replica nodes received WAL stream up to `Latest checkpoint's REDO location` and ready to be promoted.

The reason for doing it is that postgres shutdown may take infinite time because of:
1. issues with clients using logical replication
2. because it is trying to archive a backlog of WAL files

If Patroni will wait until postgres processes are terminated the downtime would be significantly higher.
Or, in other words there are always trade offs and Patroni prefers to optimize availability.

Your problem indeed could solved by properly configuring tcp keepalives.
